### PR TITLE
Sync Csc invocations from CoreCompile to XamlPreCompile

### LINKED ISSUE #9785: Error CS9137 Occurred When Building a MAUI Project Targetting to Windows
### Description

[Configuration Binding Source Generator](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-8/runtime#configuration-binding-source-generator) is introduced by .NET8.0, and it works well in my web projects. I attempt to use this generator in my MAUI Blazor App by modifying the .csproj file like:
```
<Project Sdk="Microsoft.NET.Sdk.Razor">
    <PropertyGroup>
        ...
        <EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>
        <EnableConfigurationBindingGenerator>true</EnableConfigurationBindingGenerator>
        ...
    </PropertyGroup>
    ...
</Project>
```
It does function when targetting to net8.0-android, net8.0-ios, and net8.0-maccatalyst . However, a cs9137 error occurrs when set targetting to net8.0-windows10.0.19041.0. I followed the message and added `<InterceptorsPreviewNamespaces>$(InterceptorsPreviewNamespaces);Microsoft.Extensions.Configuration.Binder.SourceGeneration</InterceptorsPreviewNamespaces>` to project file, that error still occurrs.

### Reproduction Steps

Add codes below in PropertyGroup: 
```
<EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>
<EnableConfigurationBindingGenerator>true</EnableConfigurationBindingGenerator>
<InterceptorsPreviewNamespaces>$(InterceptorsPreviewNamespaces);Microsoft.Extensions.Configuration.Binder.SourceGeneration</InterceptorsPreviewNamespaces>
```
Here is my test project [TestMauiBlazor.zip](https://github.com/dotnet/runtime/files/14392099/TestMauiBlazor.zip).

### Expected behavior

The Configuration Binding Source Generator will generate codes, and the compiler will compile the codes successfully.

### Actual behavior

The Configuration Binding Source Generator does generate codes. But the compiler throws a CS9137 error and messages &quot; The 'interceptors' experimental feature is not enabled. Add `<InterceptorsPreviewNamespaces>$(InterceptorsPreviewNamespaces);Microsoft.Extensions.Configuration.Binder.SourceGeneration</InterceptorsPreviewNamespaces>` to your project. &quot;

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

```
.NET SDK Version: 8.0.201
Operating System: Windows 10.0.22631 x64
maui-windows Version:  8.0.6/8.0.100
```

### Other information

_No response_

### PR DESCRIPTION
**Pull request by rainersigwald**
The `Csc` task invocation in `XamlPreCompile` has drifted from the one in `CoreCompile` in the Roslyn repo. I'm attempting to bring them back into sync.

This will fix #9785.

I did this in two parts to make future updates easier:

* 18671e85644dd4920bcd33f30bea73afcf530f38 is a clean copy/paste of the `Csc` & `Vbc` calls from Roslyn.

* The subsequent commit modifies them to (hopefully) have the differences we expect here.

In the future, we can update from Roslyn by syncing to caa45ead28a51cbd6322b79bec72b29102c04ce1, making a new commit of a clean update copy/paste of the task, and then merging with the subsequent edits.

### CODE REVIEW

File: src/Tasks/Microsoft.CSharp.CurrentVersion.targets [hunk 1]
```diff
@@ -234,92 +234,80 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
-         CodeAnalysisRuleSet="$(ResolvedCodeAnalysisRuleSet)"
-         CodePage="$(CodePage)"
-         DebugType="$(DebugType)"
-         DefineConstants="$(DefineConstants)"
-         DelaySign="$(DelaySign)"
-         DisabledWarnings="$(NoWarn)"
-         DisableSdkPath="$(DisableSdkPath)"
-         DocumentationFile="@(DocFileItem)"
-         EmbedAllSources="$(EmbedAllSources)"
-         EmbeddedFiles="@(EmbeddedFiles)"
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.CSharp.CurrentVersion.targets]: My reasoning here is that embedding stuff is for final outputs and isn't needed here. Should we explicitly pass false/empty though?

File: src/Tasks/Microsoft.CSharp.CurrentVersion.targets [hunk 2]
```diff
@@ -234,92 +234,80 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
-         EmitDebugInformation="$(DebugSymbols)"
-         EnvironmentVariables="$(CscEnvironment)"
-         ErrorEndLocation="$(ErrorEndLocation)"
-         ErrorLog="$(ErrorLog)"
-         ErrorReport="$(ErrorReport)"
-         Features="$(Features)"
-         InterceptorsPreviewNamespaces="$(InterceptorsPreviewNamespaces)"
-         FileAlignment="$(FileAlignment)"
-         GeneratedFilesOutputPath="$(CompilerGeneratedFilesOutputPath)"
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.CSharp.CurrentVersion.targets]: I assume this isn't relevant to the precompilation step but would like confirmation.

File: src/Tasks/Microsoft.CSharp.CurrentVersion.targets [hunk 3]
```diff
@@ -234,92 +234,80 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
-         MainEntryPoint="$(StartupObject)"
-         ModuleAssemblyName="$(ModuleAssemblyName)"
-         NoConfig="true"
-         NoLogo="$(NoLogo)"
-         NoStandardLib="$(NoCompilerStandardLib)"
-         NoWin32Manifest="$(NoWin32Manifest)"
-         Nullable="$(Nullable)"
-         Optimize="$(Optimize)"
-         Deterministic="$(Deterministic)"
-         PublicSign="$(PublicSign)"
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.CSharp.CurrentVersion.targets]: Should we keep PublicSign?

File: src/Tasks/Microsoft.CSharp.CurrentVersion.targets [hunk 4]
```diff
@@ -234,92 +234,80 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
-         OutputAssembly="@(IntermediateAssembly)"
-         OutputRefAssembly="@(IntermediateRefAssembly)"
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.CSharp.CurrentVersion.targets]: I think ref assembly outputs aren't useful here.

File: src/Tasks/Microsoft.CSharp.CurrentVersion.targets [hunk 5]
```diff
@@ -234,92 +234,80 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
-         PdbFile="$(PdbFile)"
-         Platform="$(PlatformTarget)"
-         Prefer32Bit="$(Prefer32Bit)"
-         PreferredUILang="$(PreferredUILang)"
-         ProjectName="$(MSBuildProjectName)"
-         ProvideCommandLineArgs="$(ProvideCommandLineArgs)"
-         References="@(ReferencePathWithRefAssemblies)"
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.CSharp.CurrentVersion.targets]: I don't know why this was using the old `@(ReferencePath)` before. Scared to change it but filed https://github.com/dotnet/msbuild/issues/9787.

File: src/Tasks/Microsoft.CSharp.CurrentVersion.targets [hunk 6]
```diff
@@ -234,92 +234,80 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
-         RefOnly="$(ProduceOnlyReferenceAssembly)"
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.CSharp.CurrentVersion.targets]: Definitely need to not be `RefOnly`--should we explicitly pass `false`?

File: src/Tasks/Microsoft.CSharp.CurrentVersion.targets [hunk 7]
```diff
@@ -234,92 +234,80 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
-         ReportAnalyzer="$(ReportAnalyzer)"
-         ReportIVTs="$(ReportIVTs)"
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.CSharp.CurrentVersion.targets]: Not sure what this does, should it run in this context?

File: src/Tasks/Microsoft.CSharp.CurrentVersion.targets [hunk 8]
```diff
@@ -234,92 +234,80 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
-         Resources="@(_CoreCompileResourceInputs);@(CompiledLicenseFile)"
-         ResponseFiles="$(CompilerResponseFile)"
-         RuntimeMetadataVersion="$(RuntimeMetadataVersion)"
-         SharedCompilationId="$(SharedCompilationId)"
-         SkipAnalyzers="$(_SkipAnalyzers)"
-         SkipCompilerExecution="$(SkipCompilerExecution)"
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.CSharp.CurrentVersion.targets]: Likewise `SkipCompilerExecution` seems like it wouldn't be necessary.

File: src/Tasks/Microsoft.CSharp.CurrentVersion.targets [hunk 9]
```diff
@@ -234,92 +234,80 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
-         Utf8Output="$(Utf8Output)"
-         VsSessionGuid="$(VsSessionGuid)"
-         WarningLevel="$(WarningLevel)"
-         WarningsAsErrors="$(WarningsAsErrors)"
-         WarningsNotAsErrors="$(WarningsNotAsErrors)"
-         Win32Icon="$(ApplicationIcon)"
-         Win32Manifest="$(Win32Manifest)"
-         Win32Resource="$(Win32Resource)"
-         PathMap="$(PathMap)"
-         SourceLink="$(SourceLink)">
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.CSharp.CurrentVersion.targets]: These feel like they belong only in final outputs so no need to pass them here, right?

**Review by rainersigwald — COMMENTED** [review]
@jaredpar Here are some comments on the changes I'm proposing from `CoreCompile`. Would love to get compiler team review on them please.

**Comment by jaredpar** [comment]
@RikkiGibson, @chsienki, @jjonescz FYI

File: src/Tasks/Microsoft.VisualBasic.CurrentVersion.targets [hunk 10]
```diff
@@ -270,13 +275,14 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
               Prefer32Bit="$(Prefer32Bit)"
               PreferredUILang="$(PreferredUILang)"
+              ProjectName="$(MSBuildProjectName)"
               References="@(ReferencePath)"
               RemoveIntegerChecks="$(RemoveIntegerChecks)"
               ReportAnalyzer="$(ReportAnalyzer)"
               Resources="@(_CoreCompileResourceInputs);@(CompiledLicenseFile)"
               ResponseFiles="$(CompilerResponseFile)"
               RootNamespace="$(RootNamespace)"
-              PdbFile="$(PdbFile)"
```
[the remark this answers is not in the document]
Comment (jaredpar) [reply src/Tasks/Microsoft.VisualBasic.CurrentVersion.targets:279]: Logically yes I agree there is no reason for PDB to be produced here. At the same time what is the best way to do this? I think controlling `DebugType="None"` is the more holistic approach then also pass an empty string here.

File: src/Tasks/Microsoft.CSharp.CurrentVersion.targets (hunk 1 above)
Comment (jaredpar) [reply src/Tasks/Microsoft.CSharp.CurrentVersion.targets]: I think it's important that we keep all of the properties in `<Csc` here. That way when we look at the targets file we can quickly assert whether or not properties are missing.

File: src/Tasks/Microsoft.CSharp.CurrentVersion.targets (hunk 2 above)
Comment (jaredpar) [reply src/Tasks/Microsoft.CSharp.CurrentVersion.targets]: Unfortunately everything but `FileAlignment` is necessary. The rest all impact compilation success / failure.

**Comment by RikkiGibson** [comment]
I do not think I understand what the XamlPreCompile step is doing well enough to say which arguments can be left out or not. I would lean toward including all the same arguments from the original task invocation from Roslyn as much as possible, except for those where there is a specific detrimental effect to keeping them.

**Comment by jaredpar** [comment]
@RikkiGibson part of my motivation for pinging is making you all aware that this exists. It was something I was unaware of. 

> I would lean toward including all the same arguments from the original task invocation from Roslyn as much as possible, except for those where there is a specific detrimental effect to keeping them.

Largely agree. I'm okay with a few items like disable PDB generation but overall would keep same where possible. In cases we do change arguments would use explicit values vs. implicit defaults to make it clear the item was considered.

**Comment by rainersigwald** [comment]
> Largely agree. I'm okay with a few items like disable PDB generation but overall would keep same where possible. In cases we do change arguments would use explicit values vs. implicit defaults to make it clear the item was considered.

This makes total sense and I've just tried to do so. I think it's probably best at this point to try an experimental VS with this contents and validate that we can build some WPF projects--I'll do that. Marking this as "draft" until that's done.

**Review by YuliiaKovalova — APPROVED** [review]
LGTM!

**Review by jaredpar — APPROVED** [review]

**Review by JanKrivanek — APPROVED** [review]
Looks good.
Though I do not feel 100% qualified to assess all the csc task adjustments. More details about whythose are made might be helpful.

**Pull request merged**